winml/test/model/model_tests.cpp

```diff
@@ -0,0 +1,166 @@
+  }
+};
+
+TEST_P(ModelTest, Run) {
+  LearningModel model = nullptr;
+  LearningModelDevice device = nullptr;
+  LearningModelSession session = nullptr;
+  LearningModelBinding binding = nullptr;
+  WINML_EXPECT_NO_THROW(model = LearningModel::LoadFromFilePath(m_testCase->GetModelUrl()));
+  WINML_EXPECT_NO_THROW(device = LearningModelDevice(m_deviceKind));
+  WINML_EXPECT_NO_THROW(session = LearningModelSession(model, device));
+  WINML_EXPECT_NO_THROW(binding = LearningModelBinding(session));
+  for (int i = 0; i < m_testCase->GetDataCount(); i++) {
+
+    // Load and bind inputs
+    onnxruntime::test::HeapBuffer inputHolder;
+    std::unordered_map<std::string, Ort::Value> inputFeeds;
+    WINML_EXPECT_NO_THROW(m_testCase->LoadTestData(i, inputHolder, inputFeeds, true));
+    WINML_EXPECT_NO_THROW(BindInputsFromFeed(binding, inputFeeds));
+
+    // evaluate
+    LearningModelEvaluationResult results = nullptr;
+    WINML_EXPECT_NO_THROW(results = session.Evaluate(binding, L"Testing"));
+
+    // Load expected outputs
+    onnxruntime::test::HeapBuffer outputHolder;
+    std::unordered_map<std::string, Ort::Value> outputFeeds;
+    WINML_EXPECT_NO_THROW(m_testCase->LoadTestData(i, outputHolder, outputFeeds, false));
+
+    // compare results
+    CompareEvaluationResults(results, outputFeeds);
+  }
+}
+
+void SetTestDataPath(char* testDataPath) {
+  GetEnvironmentVariableA("WINML_TEST_DATA_PATH", testDataPath, MAX_PATH);
+  if (testDataPath[0] == 0) {
+    throw std::invalid_argument("need to set environment variable WINML_TEST_DATA_PATH");
+  }
+}
+
+// This function returns the list of all test cases inside model test collateral
+std::vector<ITestCase*> GetAllTestCases() {
+  std::vector<ITestCase*> tests;
+  std::vector<std::basic_string<PATH_CHAR_TYPE>> whitelistedTestCases;
+  double perSampleTolerance = 1e-3;
+  double relativePerSampleTolerance = 1e-3;
+  std::unordered_set<std::basic_string<ORTCHAR_T>> allDisabledTests;
+  std::vector<std::basic_string<PATH_CHAR_TYPE>> dataDirs;
+  char testDataPath[MAX_PATH];
+  SetTestDataPath(testDataPath);
+
+  for (auto& p : std::filesystem::directory_iterator(testDataPath)) {
+    if (p.is_directory()) {
+      dataDirs.push_back(p.path());
+    }
+  }
+
+  WINML_EXPECT_NO_THROW(LoadTests(dataDirs, whitelistedTestCases, perSampleTolerance, relativePerSampleTolerance,
+                                  allDisabledTests,
+                                  [&tests](std::unique_ptr<ITestCase> l) {
+                                    tests.push_back(l.get());
+                                    ownedTests.push_back(std::move(l));
+                                  }));
+  return tests;
+}
+
+// This function gets the name of the test
+static std::string GetNameOfTest(const testing::TestParamInfo<ModelTest::ParamType>& info) {
+  std::string name = "";
+  auto modelPath = std::wstring(std::get<0>(info.param)->GetModelUrl());
+  auto modelPathStr = std::string(modelPath.begin(), modelPath.end());
```

Comment:
This doesn't really work (it will assign the wchar_t to the chars one by one, without any conversion, which works for ASCII characters only). Use `UTF8FromUnicode` in `_winml::Strings`

Reply:
Fixed